File: Pythoncode/34.在排序数组中查找元素的第一个和最后一个位置.py
class Solution:
    def searchRange(self, nums, target: int):
        guard = self.bin_search(nums, target, 0, len(nums)-1)
        if guard == -1:
            return [-1, -1]
        return [self.my_bin_search_left(nums, target, 0, guard), self.my_bin_search_right(nums, target, guard, len(nums)-1)]
    def bin_search(self, nums, target, head, tail):
        if head > tail:
            return -1
        else:
            mid = (head + tail) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                head = mid + 1
            else:
                tail = mid - 1
            return self.bin_search(nums, target, head, tail)


    def my_bin_search_left(self, nums, target, head, tail):
        if head == tail:
            return head
        else:
            mid = (head + tail) // 2    
            if nums[mid] == target:
                tail = mid
            else:
                head = mid + 1
            return self.my_bin_search_left(nums, target, head, tail)



    def my_bin_search_right(self, nums, target, head, tail):
        if head == tail:
            return head
        else:
            mid = (head + tail + 1) // 2    
            if nums[mid] == target:
                # if mid == head:
                #     if nums[tail] == target:
                #         return tail
                #     else:
                #         return head
                head = mid
            else:
                tail = mid - 1
            return self.my_bin_search_right(nums, target, head, tail)

File: Cppcode/34.在排序数组中查找元素的第一个和最后一个位置.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        auto a = binanrySearch(nums, target, 0, nums.size()-1);
        if (a == -1)
        {
            return vector<int> {-1, -1};
        }
        else
        {
            return vector<int> {binanrySearchLeft(nums, target, 0, a), binanrySearchRight(nums, target, a, nums.size()-1)};
        }
    }

    int binanrySearch(vector<int>& nums, int target, int head, int tail)
    {
        if (tail < head) {
            return -1;
        }
        auto mid = (head + tail) / 2;
        if (target == nums[mid])
        {
            return mid;
        }
        else if (target < nums[mid])
        {
            return binanrySearch(nums, target, head, mid-1);
        }
        else
        {
            return binanrySearch(nums, target, mid+1, tail);
        }
    }

    int binanrySearchLeft(vector<int>& nums, int target, int head, int tail)
    {
        auto mid = (head + tail) / 2;
        if (target == nums[mid] && (mid-1 == -1 || nums[mid-1] != target))
        {
            return mid;
        }
        else if (target == nums[mid])
        {
            return binanrySearchLeft(nums, target, head, mid-1);
        }
        else
        {
            return binanrySearchLeft(nums, target, mid+1, tail);
        }
    }

    int binanrySearchRight(vector<int>& nums, int target, int head, int tail)
    {
        auto mid = (head + tail) / 2;
        if (target == nums[mid] && (mid+1 == nums.size() || nums[mid+1] != target))
        {
            return mid;
        }
        else if (target == nums[mid])
        {
            return binanrySearchRight(nums, target, mid+1, tail);
        }
        else
        {
            return binanrySearchRight(nums, target, head, mid-1);
        }
    }
};